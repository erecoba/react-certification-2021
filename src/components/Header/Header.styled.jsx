import styled, { css } from 'styled-components';

const $toggleHeight = 40;
const $toggleWidth = 90;
const $borderWidth = 4;
const $toggleBtnRadius = $toggleHeight - 14;

const $bgColorNight = '#3c4145';
const $borderColorNight = '#1c1c1c';
const $toggleBtnBgColorNight = '#fff';
const $toggleBtnBorderColorNight = '#e3e3c7';

const $bgColorDay = '#9ee3fb';
const $borderColorDay = '#86c3d7';
const $toggleBtnBgColorDay = '#ffdf6d';
const $toggleBtnBorderColorDay = '#e1c348';

const $cloudBorderColor = '#d3d3d3';
const $cloudBgColor = '#fff';

const Container = styled.nav`
  position: fixed;
  top: 0;
  left: 0;

  width: 100%;
  height: 3rem;

  background-color: ${({ theme }) => theme.background};

  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;

  border-bottom: 1px solid #eaeaea;
`;

const Logo = styled.img`
  width: 30px;
  height: 30px;
  margin: 5px;
`;

const LogoContainer = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;
  cursor: pointer;
  flex-grow: 1;
`;

const LogoName = styled.h2`
  font-size: 1.3rem;
  margin: 0;
  color: ${({ theme }) => theme.label};
`;

const SearchContainer = styled.div`
  flex-grow: 1;
`;
const AvatarContainer = styled.div`
  flex-grow: 1;
  margin-right: 5px;

  display: flex;
  justify-content: flex-end;
  align-items: center;
`;

const CheckToggleWrapper = styled.div`
  margin-right: 8px;
  @keyframes starry_star {
    50% {
      background-color: rgba(255, 255, 255, 0.1);
      box-shadow: #fff 20px -3px 0 0, #fff 2px 0px 0 1px,
        rgba(255, 255, 255, 0.1) 28px 8px 0 1px, #fff 22px 14px 0 0,
        rgba(255, 255, 255, 0.1) 10px 14px 0 1.5px, #fff 2px 18px 0 1px;
    }
  }
  @keyframes bounceIn {
    0% {
      opacity: 0;
      transform: scale(0.3);
    }
    50% {
      opacity: 100;
      transform: scale(1.1);
    }
    55% {
      transform: scale(1.1);
    }
    75% {
      transform: scale(0.9);
    }
    100% {
      opacity: 100;
      transform: scale(1);
    }
  }
`;
const CheckToggle = styled.input`
  display: none;

  &,
  &:before,
  &:after {
    transition: all 250ms ease-in;
  }
  &:before,
  &:after {
    content: '';
    display: block;
  }
`;

const LabelCheckToggle = styled.label`
  display: block;
  margin: 0 auto;
  font-size: 1.4em;
  transition: all 350ms ease-in;
  &:hover {
    cursor: pointer;
  }
  /*  */
  &,
  &:before,
  &:after {
    transition: all 250ms ease-in;
  }
  &:before,
  &:after {
    content: '';
    display: block;
  }
  /*  */
  position: relative;
  height: ${$toggleHeight}px;
  width: ${$toggleWidth}px;
  border-radius: ${$toggleHeight}px;
  &:before {
    position: absolute;
    top: 2px;
    left: 4px;
    width: ${$toggleBtnRadius}px;
    height: ${$toggleBtnRadius}px;
    border-radius: 50%;
  }
  &:after {
    position: absolute;
    top: 62%;
    left: ${$toggleWidth - $toggleBtnRadius - $borderWidth * 2 - 20}px;
    z-index: 10;
    width: ${$toggleBtnRadius / 5}px;
    height: ${$toggleBtnRadius / 5}px;
    opacity: 0;
    background-color: ${$cloudBgColor};
    border-radius: 50%;
    box-shadow: ${$cloudBgColor} 0 0, ${$cloudBgColor} 3px 0, ${$cloudBgColor} 6px 0,
      ${$cloudBgColor} 9px 0, ${$cloudBgColor} 11px 0, ${$cloudBgColor} 14px 0,
      ${$cloudBgColor} 16px 0, ${$cloudBgColor} 21px -1px 0 1px,
      ${$cloudBgColor} 16px -7px 0 -2px, ${$cloudBgColor} 7px -7px 0 1px,
      ${$cloudBorderColor} 0 0 0 4px, ${$cloudBorderColor} 6px 0 0 4px,
      ${$cloudBorderColor} 11px 0 0 4px, ${$cloudBorderColor} 16px 0 0 4px,
      ${$cloudBorderColor} 21px -1px 0 5px, ${$cloudBorderColor} 16px -7px 0 1px,
      ${$cloudBorderColor} 7px -7px 0 5px;
    transition: opacity 100ms ease-in;
  }
  ${({ checked }) =>
    checked
      ? css`
          border: ${$borderWidth}px solid ${$borderColorNight};
          background-color: ${$bgColorNight};
          &:before {
            background-color: ${$toggleBtnBgColorNight};
            border: ${$borderWidth}px solid ${$toggleBtnBorderColorNight};
          }
        `
      : css`
          background-color: ${$bgColorDay};
          border: ${$borderWidth}px solid ${$borderColorDay};

          &:before {
            left: ${$toggleWidth - $toggleBtnRadius - $borderWidth * 2 - 4}px;
            background-color: ${$toggleBtnBgColorDay};
            border: ${$borderWidth}px solid ${$toggleBtnBorderColorDay};
          }
          &:after {
            opacity: 100;
            animation-name: bounceIn;
            animation-duration: 0.6s;
            animation-delay: 0.1s;
            animation-fill-mode: backwards;
            animation-timing-function: ease-in-out;
          }
        `}
`;

const LabelCheckToggleFeat = styled.span`
  &,
  &:before,
  &:after {
    transition: all 250ms ease-in;
  }
  &:before,
  &:after {
    content: '';
    display: block;
  }
  /*  */
  display: block;
  position: absolute;
  top: 9px;
  left: 52.5%;
  z-index: 20;
  width: 4px;
  height: 4px;
  border-radius: 50%;
  background-color: #fff;

  ${({ checked }) =>
    checked
      ? css`
          box-shadow: rgba(255, 255, 255, 0.1) 20px -3px 0 0,
            rgba(255, 255, 255, 0.1) 2px 0px 0 1px, #fff 28px 8px 0 1px,
            rgba(255, 255, 255, 0.1) 22px 14px 0 0, #fff 10px 14px 0 1.5px,
            rgba(255, 255, 255, 0.1) 2px 18px 0 1px;
          animation: starry_star 5s ease-in-out infinite;
          &:before {
            position: absolute;
            top: -2px;
            left: -25px;
            width: 8px;
            height: 8px;
            background-color: ${$toggleBtnBgColorNight};
            border-radius: 50%;
            border: 1px solid ${$toggleBtnBorderColorNight};
            box-shadow: ${$toggleBtnBorderColorNight} -12px 0 0 -3px,
              ${$toggleBtnBorderColorNight} -8px 10px 0 -2px;
            transform-origin: -6px 130%;
          }
        `
      : css`
          opacity: 0;
          box-shadow: rgba(255, 255, 255, 0.1) 20px -3px 0 -4px,
            rgba(255, 255, 255, 0.1) 2px 0px 0 -5px, #fff 28px 8px 0 -3px,
            rgba(255, 255, 255, 0.1) 22px 24px 0 -4px, #fff 10px 14px 0 -5.5px,
            rgba(255, 255, 255, 0.1) 2px 28px 0 -3px;
          animation: none;

          &:before {
            left: 25px;
            transform: rotate(70deg);
          }
        `};
`;

export {
  Container,
  Logo,
  LogoContainer,
  LogoName,
  SearchContainer,
  AvatarContainer,
  CheckToggleWrapper,
  CheckToggle,
  LabelCheckToggle,
  LabelCheckToggleFeat,
};
