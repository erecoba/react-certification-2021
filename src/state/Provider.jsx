import React, { createContext, useContext, useReducer } from 'react';
import youtubeReducer, { initialState as youtubeInitialState } from './reducers/youtube';

const YoutubeContext = createContext({ ...youtubeInitialState });

const useYoutubeVideo = () => {
  const context = useContext(YoutubeContext);
  if (!context) {
    throw new Error("Can't use useYoutubeVideo with out a Provider");
  }
  return context;
};

const Provider = ({ children }) => {
  const [state, dispatch] = useReducer(youtubeReducer, youtubeInitialState);

  return (
    <YoutubeContext.Provider value={{ state, dispatch }}>
      {children}
    </YoutubeContext.Provider>
  );
};

export { useYoutubeVideo };

export default Provider;
