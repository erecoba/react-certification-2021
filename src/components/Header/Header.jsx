import React, { useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';
import {
  Container,
  Logo,
  LogoContainer,
  LogoName,
  SearchContainer,
  AvatarContainer,
} from './Header.styled';

import Input from '../Input';
import Avatar from '../Avatar';
import SeachIcon from '../../ui/icons/Search';

import logo from '../../assets/img/wizeline-logo.png';

import { youtubeGetVideos, youtubeSearchVideos } from '../../state/actions/youtube';
import { useYoutubeVideo } from '../../state/Provider';

export const retrieveVideosEffect = ({ search, dispatch, hideSearcher }) => {
  if (hideSearcher) return;

  const retriveVideosFetch = async () => {
    if (search && search.trim()) {
      dispatch(await youtubeSearchVideos(search));
    } else {
      dispatch(await youtubeGetVideos());
    }
  };
  retriveVideosFetch();
};

const Header = ({ hideSearcher }) => {
  const history = useHistory();
  const [search, setSearch] = useState('');
  const { dispatch } = useYoutubeVideo();

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(() => retrieveVideosEffect({ search, dispatch, hideSearcher }), [search]);

  const handleOnChange = (event) => setSearch(event.target.value);

  return (
    <Container>
      <LogoContainer onClick={() => history.push('/')} aria-label="logo-container">
        <Logo src={logo} alt="logo" />
        <LogoName>Wizevideo</LogoName>
      </LogoContainer>

      <SearchContainer>
        {!hideSearcher && (
          <Input
            placeholder="Buscar"
            icon={<SeachIcon />}
            value={search}
            onChange={handleOnChange}
          />
        )}
      </SearchContainer>

      <AvatarContainer>
        <Avatar clickable />
      </AvatarContainer>
    </Container>
  );
};

export default Header;
