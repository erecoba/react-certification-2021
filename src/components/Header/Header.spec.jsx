import React from 'react';
import { render, act, fireEvent } from '@testing-library/react';
import 'jest-styled-components';
import { ThemeProvider } from 'styled-components';
import { BrowserRouter } from 'react-router-dom';

import Provider from '../../state/Provider';
import Header, { retrieveVideosEffect } from './Header';
import MockVideosPopular from '../../mocks/videos-popular.json';
import MockVideosSearch from '../../mocks/videos-searching.json';
import MockVideosRelated from '../../mocks/videos-related.json';
import MockVideosDetail from '../../mocks/video-detail.json';

jest.mock('../../services', () => {
  return jest.fn(() => ({
    youtube: {
      getVideos: jest.fn(() => MockVideosPopular),
      searchVideos: jest.fn(() => MockVideosSearch),
      getRelatedVideos: jest.fn(() => MockVideosRelated),
      getStatistics: jest.fn(() => MockVideosDetail),
    },
  }));
});

describe('Component - Header', () => {
  let theme;
  let props;

  const dispatch = jest.fn();
  beforeEach(() => {
    dispatch.mockReset();
    theme = {
      colors: {
        white: '#fff',
      },
    };
    props = {
      hideSearcher: false,
    };
  });

  it('should render correctly inside button', async () => {
    let wrapper;
    await act(async () => {
      wrapper = render(
        <ThemeProvider theme={theme}>
          <Provider>
            <Header {...props} />
          </Provider>
        </ThemeProvider>
      );
    });

    let fragment;
    await act(async () => {
      fragment = wrapper.asFragment(
        <ThemeProvider theme={theme}>
          <Provider>
            <Header {...props} />
          </Provider>
        </ThemeProvider>
      );
    });

    expect(fragment).toMatchSnapshot();
  });

  it('should execute youtubeSearchVideos', async () => {
    const search = 'minecraft';
    const hideSearcher = false;

    await retrieveVideosEffect({ search, dispatch, hideSearcher });

    expect(dispatch.mock.calls.length).toBe(0);
  });

  it('should execute youtubeGetVideos', async () => {
    const search = '';
    const hideSearcher = false;

    await retrieveVideosEffect({ search, dispatch, hideSearcher });

    expect(dispatch.mock.calls.length).toBe(0);
  });

  it('should do nothing while hideSearcher is true', async () => {
    const search = '';
    const hideSearcher = true;

    await retrieveVideosEffect({ search, dispatch, hideSearcher });

    expect(dispatch).toHaveBeenCalledTimes(0);
  });

  it('should click on logo to go home', async () => {
    let wrapper;
    await act(async () => {
      wrapper = render(
        <BrowserRouter>
          <ThemeProvider theme={theme}>
            <Provider>
              <Header {...props} />
            </Provider>
          </ThemeProvider>
        </BrowserRouter>
      );
    });

    const logoHref = wrapper.getByLabelText('logo-container');
    fireEvent.click(logoHref);
  });
});
