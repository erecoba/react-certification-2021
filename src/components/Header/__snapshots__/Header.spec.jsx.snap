// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Header should render correctly inside button 1`] = `
<DocumentFragment>
  .c0 {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 3rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  border-bottom: 1px solid #eaeaea;
}

.c2 {
  width: 30px;
  height: 30px;
  margin: 5px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  cursor: pointer;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c3 {
  font-size: 1.3rem;
  margin: 0;
}

.c4 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c12 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin-right: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 5px;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
}

.c7 {
  padding: 0 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c9 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin-right: 1px;
}

.c10 {
  position: relative;
}

.c11 {
  padding: 5px 10px;
  margin: 2px;
  border: 2px solid transparent;
  background-color: transparent;
  height: 2rem;
  width: 100%;
  background-clip: padding-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  font-family: 'Roobert',sans-serif;
  border-radius: 8px;
}

.c11:focus {
  outline: none;
}

.c13 {
  border-radius: 50%;
  height: 40px;
  width: 40px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.c8 {
  display: inline-block;
  width: 1.5rem;
  height: 1.5rem;
  font-size: 1.5rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c14 {
  display: inline-block;
  color: #fff;
  width: 1.5rem;
  height: 1.5rem;
  fill: #fff;
  font-size: 1.5rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

<nav
    class="c0"
  >
    <div
      aria-label="logo-container"
      class="c1"
    >
      <img
        alt="logo"
        class="c2"
        src="wizeline-logo.png"
      />
      <h2
        class="c3"
      >
        Wizevideo
      </h2>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <svg
              class="c8"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </div>
          <div
            class="c9"
          >
            <div
              class="c10"
            >
              <input
                class="c11"
                placeholder="Buscar"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c12"
    >
      <div
        class="c13"
      >
        <svg
          class="c14"
          color="white"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <circle
            cx="15.5"
            cy="9.5"
            r="1.5"
          />
          <circle
            cx="8.5"
            cy="9.5"
            r="1.5"
          />
          <circle
            cx="15.5"
            cy="9.5"
            r="1.5"
          />
          <circle
            cx="8.5"
            cy="9.5"
            r="1.5"
          />
          <path
            d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm0-2.5c2.33 0 4.32-1.45 5.12-3.5h-1.67c-.69 1.19-1.97 2-3.45 2s-2.75-.81-3.45-2H6.88c.8 2.05 2.79 3.5 5.12 3.5z"
          />
        </svg>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
